doc/Design.md

```diff
@@ -143,3 +159,142 @@ dereferencing. Unlike shared buffers, which can only be a buffer type in a
 capsule, granted memory can be defined as any type. Therefore, processes cannot
 access this memory since doing so might violate type-safety.
 
+## In-Kernel Design Principles
+
+To help meet Tock's goals, encourage portability across hardware, and ensure a
+sustainable operating system, several design principles have emerged over time
+for the Tock kernel. These are general principles that new contributions to the
+kernel should try to uphold. However, these principles have been informed by
+Tock's development, and will likely continue to evolve as Tock and the Rust
+ecosystem evolve.
+
+### Role of HILs
+
+Generally, the Tock kernel is structured into three layers:
+
+1. Chip-specific drivers: these typically live in a crate in the
+   `chips` subdirectory, or an equivalent crate in an different repository
+   (e.g. the Titan port is out of tree but its `h1b` create is the
+   equivalent here). These drivers have implementations that are specific
+   to the hardware of a particular microcontroller. Ideally,
+   their implementation is fairly simple, and they merely adhere to a
+   common interface (a HIL). That's not always the case, but that's
+   the ideal.
+
+2. Chip-agnostic, portable, peripheral drivers and subsystems. These
+   typically live in the `capsules` crate. These includes things like
+   the virtual alarms and virtual I2C stack, as well as drivers for
+   hardware peripherals not on the chip itself (e.g. sensors, radios,
+   etc). These drivers typically rely on the chip-specific drivers
+   through the HILs.
+
+3. System call drivers, also typically found in the `capsules`
+   crate. These are the drivers that implement a particular part of
+   the system call interfaces, and are often even more abstracted from
+   the hardware than (2) - for example, the temperature sensor system
+   call driver can use any temperature sensor, including several
+   implemented as portable peripheral drivers.
+
+
+   The system call interface is another point of
+   standardization that can be implemented in various ways. So it’s
+   perfectly reasonable to have several implementations of the same
+   system call interface that use completely different hardware
+   stacks, and therefore HILs and chip-specific drivers (e.g. a
+   console driver that operates over USB might just be implemented as
+   a different system call driver that implements the same system
+   calls, rather than trying to fit USB into the UART HIL).
+
+The connective tissue between layers are the HILs. The HIL interfaces
+are portable interfaces that are implemented in a non-portable way.
+
+The choice of particular HIL interfaces is pretty important, and we
+have some general principles we follow:
+
+1. HIL implementations get to assume this HIL is the only way the device will be
```

Comment:
I don't quite understand this. Is a HIL a trait or a file? E.g., I'd expect an ADC to be able to support single samples as well as continuous samples; a virtualization layer would queue the requests. I think that the design principle is that the behavior of a trait should be isolated to that trait. I.e., if continuous sampling and single sampling are different traits, then single sample should not return EBUSY if there is a continuous sample going on. However, if they are the same trait then it could return EBUSY.

What is this getting at? Is it that we don't have independent/different HILs?

Reply:
Does

> For example, two separate HIL traits could exist for the ADC: one that only provides single samples and one that allows for repeated, periodic samples. If the two interfaces provide access to the same hardware and are used simultaneously, likely one would corrupt the settings of the other, and at least one of the interfaces would not function correctly.

help clarify? I thought it would be useful to include a concrete example here.

Reply:
I don't think this helps -- is the HIL the full set of traits in a file or a single trait? The definition given makes it sound like the latter, but I think we usually mean the former. I also disagree with the example: it implies that a single ADC implementation can't coordinate between adc::Adc and adc::AdcHighSpeed. How about:

"Because of their importance, the interfaces between these layers are a key part of Tock's design and implementation. These interfaces are called Tock's Hardware Interface Layer, or HIL. A HIL is a portable collections of Rust traits that can be implemented in either a portable or a non-portable way. An example of a non-portable implementation of a HIL is an Alarm that is implemented in terms of counter and compare registers of a specific chip, while an example of a portable implementation is a virtualization layer that multiplexes multiple Alarms top of a single underlying Alarm. 

A HIL consists of one or more Rust traits that are intended to be used together. In some cases, implementations may only implement a subset of a HIL's traits. For example the analog-to-digital conversion HIL may have traits both for single and streams of samples. A particular implementation may only support single samples and so not implement the streaming traits.

The choice of particular HIL interfaces is pretty important, and we have some general principles we follow:

1. A HIL implementation may assume it is the only way the device will be used. As a result, Tock tries to avoid having more than one HIL for a particular service or abstraction, because it will not, in general, be possible for the kernel to support simultaneously using different HILs for the same device. For example, suppose there were two different HILs for a UART with slightly different APIs. The chip-specific implementation of each one will need to read and write hardware registers and handle interrupts, so they cannot exist simultaneously. By allowing a HIL to assume it is the only way the device will be used, Tock allows HILs to precisely define their semantics without having to worry about potential future conflicts or use cases."

Reply:
This also seems weird to have as the first bullet. It seems like a secondary point to the more important one (currently bullet 2), that they should be general.